lib/serial_terminal.pm

```diff
@@ -96,4 +101,105 @@ sub serial_term_prompt {
     return $serial_term_prompt;
 }
 
+
+=head2 download_file
+
+  download_file($src, $dst [, force => $force][, chunk_size => $cz][, chunk_retry => $cr])
+
+Download a file from worker to SUT using the current serial terminal.
+The file is split into chunks C<chunk_size> and each chunk is verified with
+checksum. If a chunk fail we retry it C<chunk_retry> times, before give up.
+To override destination use C<force>.
+This function die on any failure.
+=cut
+sub download_file {
+    my ($src, $dst, %opts) = @_;
+    $opts{chunk_size}  //= 1024 * 2;
+    $opts{chunk_retry} //= 16;
+    $opts{force}       //= 0;
+    record_info('Download file', "From worker ($src) to SUT ($dst)");
+    die("Relative path is forbidden - '$src'") if $src =~ m'/\.\.|\.\./';
+    $src =~ s'^/+'';
+    if ($src =~ m'data/') {
+        $src = $bmwqemu::vars{CASEDIR} . '/' . $src;
+    } else {
+        $src = $bmwqemu::vars{ASSETDIR} . '/' . $src;
+    }
+    die("File $dst already exists on SUT") if (!$opts{force} && script_run("test -f $dst", undef, quiet => 1) == 0);
+    die("File $src not found on worker") unless (-f $src);
+    my $tmpdir = script_output('mktemp -d', undef, quiet => 1);
+    assert_script_run("test -d $tmpdir", quiet => 1);
+
+    open(my $fh, '<:raw', $src) or die "Could not open file '$src' $!";
+    my $result_file = $tmpdir . '/result';
+    my $tmpfile     = $tmpdir . '/chunk';
+    my $cnt         = 0;
+    while (my $read = read($fh, my $chunk, $opts{chunk_size})) {
+        my $b64   = b64_encode($chunk);
+        my $tries = $opts{chunk_retry};
+        my $sha1  = sha1_sum($b64);
+        $cnt += 1;
+        do {
+            die("Failed to transfer chunk[$cnt] of file $src") if ($tries-- < 0);
+            script_output("cat > $tmpfile << 'EOT'\n" . $b64 . "EOT", undef, quiet => 1);
+        } while ($sha1 ne script_output("sha1sum $tmpfile | cut -d ' ' -f 1", undef, quiet => 1));
+        assert_script_run("base64 -d $tmpfile >> $result_file", quiet => 1);
+        assert_script_run("rm $tmpfile",                        quiet => 1);
+    }
+    close($fh);
+    my $sha1_remote = script_output("sha1sum $result_file | cut -d ' ' -f 1", undef, quiet => 1);
+    my $sha1        = trim(`sha1sum $src | cut -d ' ' -f 1`);
```

Comment:
this is just because it hapens already on the worker and not in the sut right?

Reply:
Yes. The file is transferred chunk by chunk and this gets the sha1sum of the _local_ file before it get's moved to final destination.